File: outstanding-amount-stale-after-refund/node/detect-stale-outstanding.js
/**
 * Flag Medusa v2 orders whose outstanding_amount stopped updating after the
 * first refund. outstanding_amount is a derived field on the order's summary,
 * computed by the totals module from order_transaction rows, not a value that
 * gets decremented directly. The first refund on an order inserts a new
 * transaction row and the summary recomputes correctly, but a second
 * refundPaymentsWorkflow run on the same order or payment does not insert
 * another row (see medusajs/medusa#11481), so the summary is never recomputed
 * again and outstanding_amount freezes while the payment provider keeps
 * processing more refunds. There is no safe PATCH for this field, so this
 * script only flags the divergence for a human to reconcile. It never calls
 * the refund endpoint again. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/outstanding-amount-stale-after-refund/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const EPSILON = 0.01;

const ORDER_FIELDS =
  "id,display_id,*summary,*payment_collections," +
  "*payment_collections.payments,*payment_collections.payments.refunds";

export function detectStaleOutstanding(order) {
  // Pure: no I/O. order = { total, captures: [{amount}],
  // refunds: [{id, amount, created_at}], reportedOutstanding }
  const captures = order.captures || [];
  const refunds = order.refunds || [];

  const trueOutstanding =
    order.total -
    captures.reduce((sum, c) => sum + c.amount, 0) +
    refunds.reduce((sum, r) => sum + r.amount, 0);
  const refundCount = refunds.length;
  const delta = order.reportedOutstanding - trueOutstanding;
  const affected = refundCount > 1 && Math.abs(delta) > EPSILON;

  return {
    affected,
    trueOutstanding,
    reportedOutstanding: order.reportedOutstanding,
    delta,
    refundCount,
  };
}

function toDecisionInput(rawOrder) {
  const payments = (rawOrder.payment_collections || []).flatMap(
    (collection) => collection.payments || []
  );
  const captures = payments
    .filter((p) => p.captured_at)
    .map((p) => ({ amount: p.amount || 0 }));
  const refunds = payments.flatMap((p) =>
    (p.refunds || []).map((r) => ({ id: r.id, amount: r.amount || 0, created_at: r.created_at }))
  );
  const summary = rawOrder.summary || {};

  return {
    id: rawOrder.id,
    displayId: rawOrder.display_id,
    total: summary.raw_current_order_total ?? rawOrder.total ?? 0,
    captures,
    refunds,
    reportedOutstanding: summary.outstanding_amount || 0,
  };
}

async function getAdminToken() {
  const res = await fetch(`${BACKEND_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: ADMIN_EMAIL, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function adminGet(token, path, params = {}) {
  const url = new URL(`${BACKEND_URL}${path}`);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${path} ${res.status}`);
  return res.json();
}

async function listOrdersWithRefunds(token) {
  const orders = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const data = await adminGet(token, "/admin/orders", {
      fields: ORDER_FIELDS,
      limit,
      offset,
    });
    orders.push(...data.orders);
    offset += limit;
    if (offset >= data.count) return orders;
  }
}

export async function run() {
  const token = await getAdminToken();
  const rawOrders = await listOrdersWithRefunds(token);

  let flagged = 0;
  for (const rawOrder of rawOrders) {
    const decisionInput = toDecisionInput(rawOrder);
    const outcome = detectStaleOutstanding(decisionInput);
    if (!outcome.affected) continue;

    flagged++;
    console.warn(
      `Order ${decisionInput.displayId || decisionInput.id} stale outstanding_amount: ` +
        `reported=${outcome.reportedOutstanding} true=${outcome.trueOutstanding} ` +
        `delta=${outcome.delta} refund_count=${outcome.refundCount} ` +
        `refunds=${JSON.stringify(decisionInput.refunds)}. ` +
        `${DRY_RUN ? "would flag for review" : "flagging for review"}`
    );
  }

  console.log(
    `Done. ${flagged} order(s) with a stale outstanding_amount flagged for manual reconciliation. ` +
      `No orders were written to.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: outstanding-amount-stale-after-refund/python/detect_stale_outstanding.py
"""Flag Medusa v2 orders whose outstanding_amount stopped updating after the
first refund. outstanding_amount is a derived field on the order's summary,
computed by the totals module from order_transaction rows, not a value that
gets decremented directly. The first refund on an order inserts a new
transaction row and the summary recomputes correctly, but a second
refundPaymentsWorkflow run on the same order or payment does not insert
another row (see medusajs/medusa#11481), so the summary is never recomputed
again and outstanding_amount freezes while the payment provider keeps
processing more refunds. There is no safe PATCH for this field, so this
script only flags the divergence for a human to reconcile. It never calls
the refund endpoint again. Safe to run again and again.

Guide: https://www.allanninal.dev/medusa/outstanding-amount-stale-after-refund/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_stale_outstanding")

BACKEND_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
ADMIN_EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
ADMIN_PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

EPSILON = 0.01

ORDER_FIELDS = (
    "id,display_id,*summary,*payment_collections,"
    "*payment_collections.payments,*payment_collections.payments.refunds"
)


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def admin_get(token, path, params=None):
    r = requests.get(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def detect_stale_outstanding(order):
    """Pure decision function. No I/O.

    order: {
      "total": float,
      "captures": [{"amount": float}],
      "refunds": [{"id": str, "amount": float, "created_at": str}],
      "reportedOutstanding": float,
    }

    Returns {"affected", "trueOutstanding", "reportedOutstanding", "delta",
             "refundCount"}.
    """
    captures = order.get("captures") or []
    refunds = order.get("refunds") or []

    true_outstanding = (
        order["total"]
        - sum(c["amount"] for c in captures)
        + sum(r["amount"] for r in refunds)
    )
    refund_count = len(refunds)
    delta = order["reportedOutstanding"] - true_outstanding
    affected = refund_count > 1 and abs(delta) > EPSILON

    return {
        "affected": affected,
        "trueOutstanding": true_outstanding,
        "reportedOutstanding": order["reportedOutstanding"],
        "delta": delta,
        "refundCount": refund_count,
    }


def to_decision_input(raw_order):
    payments = [
        payment
        for collection in (raw_order.get("payment_collections") or [])
        for payment in (collection.get("payments") or [])
    ]
    captures = [{"amount": p.get("amount", 0)} for p in payments if p.get("captured_at")]
    refunds = [
        {"id": r.get("id"), "amount": r.get("amount", 0), "created_at": r.get("created_at")}
        for p in payments
        for r in (p.get("refunds") or [])
    ]
    summary = raw_order.get("summary") or {}

    return {
        "id": raw_order.get("id"),
        "displayId": raw_order.get("display_id"),
        "total": summary.get("raw_current_order_total", raw_order.get("total", 0)),
        "captures": captures,
        "refunds": refunds,
        "reportedOutstanding": summary.get("outstanding_amount", 0),
    }


def list_orders_with_refunds(token):
    orders = []
    offset = 0
    limit = 100
    while True:
        data = admin_get(token, "/admin/orders", {
            "fields": ORDER_FIELDS,
            "limit": limit,
            "offset": offset,
        })
        orders.extend(data["orders"])
        offset += limit
        if offset >= data["count"]:
            return orders


def run():
    token = get_admin_token()
    raw_orders = list_orders_with_refunds(token)

    flagged = 0
    for raw_order in raw_orders:
        decision_input = to_decision_input(raw_order)
        outcome = detect_stale_outstanding(decision_input)
        if not outcome["affected"]:
            continue

        flagged += 1
        log.warning(
            "Order %s stale outstanding_amount: reported=%s true=%s delta=%s "
            "refund_count=%d refunds=%s. %s",
            decision_input["displayId"] or decision_input["id"],
            outcome["reportedOutstanding"], outcome["trueOutstanding"], outcome["delta"],
            outcome["refundCount"],
            [(r["id"], r["amount"], r["created_at"]) for r in decision_input["refunds"]],
            "would flag for review" if DRY_RUN else "flagging for review",
        )

    log.info(
        "Done. %d order(s) with a stale outstanding_amount flagged for manual reconciliation. "
        "No orders were written to.",
        flagged,
    )


if __name__ == "__main__":
    run()
